examples/small-coalton-programs/src/brainfold.lisp

```diff
@@ -0,0 +1,308 @@
+    "Converts a UFix into an ASCII string."
+    (into (make-list (unwrap (char:code-char ascii-value)))))
+
+  (declare print (State -> State))
+  (define (print s)
+    "Prints the current bf-cell."
+    (cell:write! (state-print-buffer s)
+                 (str:concat
+                  (cell:read (state-print-buffer s))
+                  (find-ascii
+                   (fromint (abs (value-at-pointer s))))))
+    s)
+
+  ;;
+  ;; Taking Input (,)
+  ;; 
+  ;; Currently takes individual characters one at a time as prompted
+
+  (define (character-prompt)
+    "A prompt for obtaining one character as input."
+    (Lisp Char ()
+      (cl:format cl:*query-io* "Input a character: ")
+      (cl:finish-output cl:*query-io*)
+      (cl:read-char cl:*query-io*)))
+
+  (declare take-input (State -> State))
+  (define (take-input s)
+    "Takes and stores a character at the input-index as an ascii code at
+the pointer."
+    (vec:set! (cell:read (state-pointer s))
+              (into (char:char-code (character-prompt)))
+              (state-memory s))
+    s))
+
+
+;;;
+;;; Parsing/Lexing
+;;;
+
+(coalton-toplevel
+
+  (define-type Cmd
+    BFRight
+    BFLeft
+    BFPlus
+    BFMinus
+    BFPrint
+    BFInput
+    (BFLoop (Vector Cmd)))
+
+  (declare parse (String -> (Vector Cmd)))
+  (define (parse input-string)
+    "Parses a Brainfold instruction string, returns a Vector of Brainfold Commands."
+    (let cmds = (vec:new))
+    (let vecs = (vec:new))
+    (let ((parser (fn (input-string v)
+                    (let ((head-tail (str:split 1 input-string)))
+                      (match (fst head-tail)
+                        ("" cmds)
+                        (">"
+                         (vec:push! BFRight v)
+                         (parser (snd head-tail) v))
+                        ("<"
+                         (vec:push! BFLeft v)
+                         (parser (snd head-tail) v))
+                        ("+"
+                         (vec:push! BFPlus v)
+                         (parser (snd head-tail) v))
+                        ("-"
+                         (vec:push! BFMinus v)
+                         (parser (snd head-tail) v))
+                        ("."
+                         (vec:push! BFPrint v)
+                         (parser (snd head-tail) v))
+                        (","
+                         (vec:push! BFInput v)
+                         (parser (snd head-tail) v))
+                        ("["
+                         (vec:push! v vecs)
+                         (parser (snd head-tail) (vec:new)))
+                        ("]"
+                         (vec:push! (BFLoop v) (unwrap (vec:last vecs)))
+                         (parser (snd head-tail) (unwrap (vec:pop! vecs))))
+                        (_ (parser (snd head-tail) v)))))))
+      (parser input-string cmds))))
+
+
+;;;
+;;; Evaluation
+;;;
+
+(coalton-toplevel
+  
+  (declare exec (Cmd -> State -> State))
```

Comment:
another way to write this that emphasizes Cmds corresponding to State transitions:

```
  (declare exec (Cmd -> (State -> State))) ;the extra parens are redundant
  (define (exec cmd)
    "Executes a given bf command string."
    (match cmd
      ((BFRight)  move-right)
      ((BFLeft)   move-left)
      ((BFPlus)   incr)
      ((BFMinus)  decr)
      ((BFPrint)  print)
      ((BFInput)  take-input)
      ((BFLoop v) (exec-loop v))))
```

I'm not suggesting this style is preferable.